Add embedding models configurable, from both Xenova and TEI

mikelfried opened the pull request:
### Embedding models configurable, from both Xenova and TEI

Enable to configure the embedding model for the websearch. it allows running the model both locally and on hosted machines utilizing GPUs.

* there can be multiple embedding models, the first is the default, and for each text generation model you can set their own specific embedding model (could be language dependent).
* support for both Xenova (currently supported in chat-ui) and [TEI](https://github.com/huggingface/text-embeddings-inference) (Text Embeddings Inference)

#### Work in progress
@mishig25 in #641 PR changes the current embedding model code, so will be need to be tweaked according to those changes. This feature will be helpfull to get faster and larger embedding models for the pdf chunks (when tei hosted on gpu), and can be language specific.

#### current limitations
* when using TEI, there is env varaibles there `MAX_CLIENT_BATCH_SIZE`, `MAX_BATCH_TOKENS` that limits the size of batch, so need to take it into account. (I should make `/info` request to the endpoint first and then send batches accordingly)

#### Status
Currently works, will fix the limitations soon.

please share your opinion 🙂

nsarrazin reviewed (commented):
Hi, thanks for the contribution! Looks like a great PR, I like the way you handled the embeddings coming from different sources, should be quite easy to extend in the future.

I was wondering why you chose [to specify the embedding at the model level](https://github.com/huggingface/chat-ui/pull/646/files#diff-ac854d5672af10370323e57d41c389221362fd685d219fc5b74c6ff0d23fc6e8R69) ? 

For most use cases, I imagine that specifying a single embedding config for the entire app should be enough no? Not 100% sure on this, maybe there are use cases where you need to support multiple embeddings, let me know!

mikelfried commented:
> Hi, thanks for the contribution! Looks like a great PR, I like the way you handled the embeddings coming from different sources, should be quite easy to extend in the future.
> 
> I was wondering why you chose [to specify the embedding at the model level](https://github.com/huggingface/chat-ui/pull/646/files#diff-ac854d5672af10370323e57d41c389221362fd685d219fc5b74c6ff0d23fc6e8R69) ?
> 
> For most use cases, I imagine that specifying a single embedding config for the entire app should be enough no? Not 100% sure on this, maybe there are use cases where you need to support multiple embeddings, let me know!

Hi, thanks for the response,
Without specifically defining embeddingModelName, the first embedding model will be used. The embeddingModelName will be used when the user has one English llm and another Korean (or any language) so will want to use other embedding model for none English,
Or there will be code llm and an embedding model that especially good for codes. This is my reasoning behind the optional embeddingModelName parameter.

nsarrazin commented:
Of course, then it would make a lot of sense to have different embeddings, I didn't think about those use cases :grin:

mikelfried commented:
Updated with automatic batching for TEI endpoints, using /info route to calculate max batch size.

Fixed a bug: when using Web Search and the llm finish outputing the web search box and the sources disappeared, and got visible again when refreshing the page.

mikelfried commented:
Now supports models with necessary prefixes, such as https://huggingface.co/intfloat/multilingual-e5-large#faq, with `preQuery` and `prePassage` optional parameters.

mishig25 commented on src/routes/conversation/[id]/+page.svelte:
```diff
@@ -138,6 +138,7 @@
 			// eslint-disable-next-line no-undef
 			const encoder = new TextDecoderStream();
 			const reader = response?.body?.pipeThrough(encoder).getReader();
+			let importantUpdates: MessageUpdate[] = [];
```
what;s the purpose of this variable `importantUpdates` ?

mikelfried replied:
I did it to solved the bug: when using Web Search and the llm finish outputing the web search box and the sources disappeared, and got visible again when refreshing the page.

the problem was that the updates of the current web search wasn’t saved into the updates in message inside the messages array, so I kept all the relevant updates (all without stream) in an array, and then saved it into the last message updates in the messages array.

mikelfried replied:
I will postpone the bug fix into a different PR, it isn’t related to this feature.

mishig25 commented:
Another small comment is: instead of calling `xeonva`, we should call that type of model `transformersjs`

mishig25 commented on src/lib/server/embeddingEndpoints/tei/embeddingEndpoints.ts:
```diff
@@ -0,0 +1,65 @@
+import { chunk } from "$lib/utils/chunk";
+
+export const embeddingEndpointTeiParametersSchema = z.object({
+	weight: z.number().int().positive().default(1),
```
how is `weight` parameter used?

mikelfried replied:
It is used exactly like the endpoint of the llms, send requests to each endpoint in proportion to its weights

mishig25 commented on src/lib/server/embeddingEndpoints/embeddingEndpoints.ts:
```diff
@@ -0,0 +1,38 @@
+) => EmbeddingEndpoint;
+
+// list of all endpoint generators
+export const embeddingEndpoints = {
+	tei: embeddingEndpointTei,
+	xenova: embeddingEndpointXenova,
```
I think you can type it even stronger here. It can look something like:
```ts
const EMBEDDING_ENDPOINT_TYPES = ["tei", "xenova"];
type EmbeddingEndpointType = typeof EMBEDDING_ENDPOINTS[number];
export const embeddingEndpoints: Record<EmbeddingEndpointType, ...> = {
	tei: embeddingEndpointTei,
	xenova: embeddingEndpointXenova,
}

```

mishig25 commented on src/lib/server/embeddingEndpoints/embeddingEndpoints.ts:
```diff
@@ -0,0 +1,38 @@
+	embeddingEndpointTeiParametersSchema,
+} from "./tei/embeddingEndpoints";
+import { z } from "zod";
+import embeddingEndpointXenova, {
+	embeddingEndpointXenovaParametersSchema,
+} from "./xenova/embeddingEndpoints";
```
```suggestion
import {
    embeddingEndpointXenova,
	embeddingEndpointXenovaParametersSchema,
} from "./xenova/embeddingEndpoints";
```

to match as above 
```ts
import {
	embeddingEndpointTei,
	embeddingEndpointTeiParametersSchema,
} from "./tei/embeddingEndpoints";
```

mishig25 commented on src/lib/server/embeddingEndpoints/tei/embeddingEndpoints.ts:
```diff
@@ -0,0 +1,65 @@
+			})
+		);
+
+		const allEmbeddings = batchesResults.flatMap((embeddings) => embeddings);
```
```suggestion
		const allEmbeddings = batchesResults.flat();
```

haven't checked. But is it possible to just call `flat()` ?

mishig25 commented on src/lib/server/embeddingEndpoints/embeddingEndpoints.ts:
```diff
@@ -0,0 +1,38 @@
+import {
```
it makes sense to put this file inside https://github.com/huggingface/chat-ui/tree/main/src/lib/types folder

mishig25 commented:
Overall, great work ! Looking forward to merging it 🚀

mikelfried commented:
Hi 😀, @mishig25 and @nsarrazin, It is working now, I fixed your past reviews.

mishig25 commented:
@mikelfried reviewing it now. Apologies for delay as I was off for vacations :)

mishig25 commented on README.md:
```diff
@@ -78,10 +78,46 @@ Chat UI features a powerful Web Search feature. It works by:
 4. From these embeddings, find the ones that are closest to the user query using a vector similarity search. Specifically, we use `inner product` distance.
 5. Get the corresponding texts to those closest embeddings and perform [Retrieval-Augmented Generation](https://huggingface.co/papers/2005.11401) (i.e. expand user prompt by adding those texts so that an LLM can use this information).
 
+### Text Embedding Models
```
```suggestion
## Text Embedding Models
```

it can be heading 2 and you can add this section to the toctree list [here](https://github.com/huggingface/chat-ui/blob/feature/assistants/README.md#L22)

mishig25 commented on README.md:
```diff
@@ -78,10 +78,46 @@ Chat UI features a powerful Web Search feature. It works by:
 
+### Text Embedding Models
+
+By default (for backward compatibility) when not defining TEXT_EMBEDDING_MODELS environment variable It will use [transformers.js](https://huggingface.co/docs/transformers.js), specifically, [Xenova/gte-small](https://huggingface.co/Xenova/gte-small) model.
```
```suggestion
By default (for backward compatibility), when `TEXT_EMBEDDING_MODELS` environment variable is not defined, [transformers.js](https://huggingface.co/docs/transformers.js) embedding models will be used for embedding tasks, specifically, [Xenova/gte-small](https://huggingface.co/Xenova/gte-small) model.
```

mishig25 commented on README.md:
```diff
@@ -78,10 +78,46 @@ Chat UI features a powerful Web Search feature. It works by:
+
+By default (for backward compatibility) when not defining TEXT_EMBEDDING_MODELS environment variable It will use [transformers.js](https://huggingface.co/docs/transformers.js), specifically, [Xenova/gte-small](https://huggingface.co/Xenova/gte-small) model.
+
+You can customize the embedding model by setting TEXT_EMBEDDING_MODELS in your `.env.local`, for example
```
```suggestion
You can customize the embedding model by setting `TEXT_EMBEDDING_MODELS` in your `.env.local` file. For example:
```

mishig25 commented on README.md:
````diff
@@ -78,10 +78,46 @@ Chat UI features a powerful Web Search feature. It works by:
+]`
+```
+
+The required fields are `name`, `maxSequenceLength` and `endpoints`.
+It supports [transformers.js](https://huggingface.co/docs/transformers.js) and [TEI](https://github.com/huggingface/text-embeddings-inference), transformers.js model run locally, and TEI models run in a different environment. each `endpoints` provided supports a `weight` parameter which will be used to determine the probability of requesting a particular endpoint.
````
```suggestion
The required fields are `name`, `maxSequenceLength` and `endpoints`.
Supported text embedding backends are: [`transformers.js`](https://huggingface.co/docs/transformers.js) and [`TEI`](https://github.com/huggingface/text-embeddings-inference). `transformers.js` models run locally as part of `chat-ui`, whereas `TEI` models run in a different environment & accessed through an API endpoint.
```

mishig25 commented on README.md:
```diff
@@ -78,10 +78,46 @@ Chat UI features a powerful Web Search feature. It works by:
+The required fields are `name`, `maxSequenceLength` and `endpoints`.
+It supports [transformers.js](https://huggingface.co/docs/transformers.js) and [TEI](https://github.com/huggingface/text-embeddings-inference), transformers.js model run locally, and TEI models run in a different environment. each `endpoints` provided supports a `weight` parameter which will be used to determine the probability of requesting a particular endpoint.
+
+When defining more than one embedding model, the first will be used by default, and the others will only be used on LLM's which configured `embeddingModelName` to the name of the model.
```
```suggestion
When more than one embedding models are supplied in `.env.local` file, the first will be used by default, and the others will only be used on LLM's which configured `embeddingModelName` to the name of the model.
```

mishig25 commented on src/lib/types/EmbeddingEndpoints.ts:
```diff
@@ -0,0 +1,39 @@
+import { z } from "zod";
```
small nit: this file should be named from `EmbeddingEndpoints.ts` to `EmbeddingEndpoint.ts` (without the plural `s`)

mishig25 commented on src/lib/types/EmbeddingEndpoints.ts:
```diff
@@ -0,0 +1,39 @@
+}
+
+// type signature for the endpoint
+export type EmbeddingEndpoint = (params: EmbeddingEndpointParameters) => Promise<number[][]>;
```
You can extract `number[][]` into type, such as `type Embedding = number[][]` or something like that

mishig25 commented on src/lib/server/embeddingEndpoints/tei/embeddingEndpoints.ts:
```diff
@@ -0,0 +1,63 @@
+		},
+	});
+
+	const info = await response.json();
+
+	return info;
```
```suggestion
	const json = await response.json();
	return json;
```

mishig25 commented on src/lib/server/models.ts:
```diff
@@ -66,6 +67,7 @@ const modelConfig = z.object({
 		.optional(),
 	multimodal: z.boolean().default(false),
 	unlisted: z.boolean().default(false),
+	embeddingModelName: validateEmbeddingModelByName(embeddingModels).optional(),
```
```suggestion
	embeddingModel: validateEmbeddingModelByName(embeddingModels).optional(),
```

I think this field can be just renmaed to `embeddingModel`. In this case, you'd need to update the references that were using `embeddingModelName` before

mishig25 replied:
here as well https://github.com/huggingface/chat-ui/blob/feature/assistants/src/routes/conversation/%2Bserver.ts#L40-L41

mishig25 commented on src/routes/conversation/+server.ts:
```diff
@@ -22,6 +23,7 @@ export const POST: RequestHandler = async ({ locals, request }) => {
 		.parse(JSON.parse(body));
 
 	let preprompt = values.preprompt;
+	let embeddingModelName: string;
```
like above https://github.com/huggingface/chat-ui/pull/646#discussion_r1442941263

mishig25 commented on src/lib/server/embeddingEndpoints/tei/embeddingEndpoints.ts:
```diff
@@ -0,0 +1,63 @@
+			batchesInputs.map(async (batchInputs) => {
+				const response = await fetch(`${origin}/embed`, {
+					method: "POST",
+					headers: {
```
in TEI parameters, could yiu define an optional authorization field that will look like:
```
authorization:{
    type: string // can be one of Bearer, Basic, OAuth, etc.
    accessToken: string // the token
}
```

and if `authorization` is defined, then add it to headers like
`headers.append('Authorization', `${authorization.type} ${authorization.accessToken}`);`

mishig25 replied:
applies to `getModelInfoByUrl` above as well

mikelfried commented:
Hi @mishig25, thanks for the review, fixed it.

mishig25 commented:
LGTM ! Asking a review from @nsarrazin

mishig25 commented:
@nsarrazin you can more attention to embedding endpoints files: [types/EmbeddingEndpoints.ts](https://github.com/mikelfried/chat-ui/blob/embedding_models/src/lib/types/EmbeddingEndpoints.ts), [transformersjs/embeddingEndpoints.ts](https://github.com/mikelfried/chat-ui/blob/embedding_models/src/lib/server/embeddingEndpoints/transformersjs/embeddingEndpoints.ts), [tei/embeddingEndpoints.ts](https://github.com/mikelfried/chat-ui/blob/embedding_models/src/lib/server/embeddingEndpoints/tei/embeddingEndpoints.ts) as they resemble the structure a lot of https://github.com/huggingface/chat-ui/pull/541

mishig25 commented on README.md:
```diff
@@ -425,6 +462,44 @@ If you're using a certificate signed by a private CA, you will also need to add
+    "name": "Xenova/gte-small",
+    "maxSequenceLength": 512,
+    "endpoints": [
+      {"type": "xenova"}
```
```suggestion
      {"type": "transformersjs"}
```

mishig25 commented on README.md:
```diff
@@ -425,6 +462,44 @@ If you're using a certificate signed by a private CA, you will also need to add
+      {"type": "xenova"}
+    ]
+  },
+  {
```
could you add [dummy authorization field ](https://github.com/mikelfried/chat-ui/blob/embedding_models/src/lib/server/embeddingEndpoints/tei/embeddingEndpoints.ts#L51-L52) example here?

mishig25 commented:
@mikelfried could you run the formatter again to pass the CI ?

mishig25 commented:
@mikelfried @nsarrazin actually I think we need one more config/setting: `embedding endpoint for task`.
For example, we would very likely use:
1. transformersjs for web-search
2. tei for https://github.com/huggingface/chat-ui/pull/641

Therefore, we would need a setting to specify by task as well. Wdyt?

nsarrazin commented:
I can see the use case, but not sure what would be the best way to structure it in the config file. Maybe we could have three vars inside of the model:

`embeddingModel` which is the default plus `searchEmbeddingModel` and `pdfEmbeddingModel` to (optionally) override the base `embeddingModel`  ? not sure about this, open to alternatives

mishig25 commented:
> embeddingModel which is the default plus searchEmbeddingModel and pdfEmbeddingModel to (optionally) override the base embeddingModel ? not sure about this, open to alternatives

this option sounds good to me

mishig25 reviewed (approved):
from my side, this PR, LGTM !

nsarrazin reviewed (approved):
This is working great, quite happy with it. I tested locally with the huggingchat config and it works out of the box. Thanks for the great contribution @mikelfried and I think we can merge this now!

We can add the extra vars in the PDF PR @mishig25, thanks for the detailed review as well!

mishig25 commented:
one sec, don't merge yet

mishig25 commented:
Ready to merge ! @nsarrazin 

@mikelfried great job again !

nsarrazin commented:
Testing one last time that it doesn't break huggingchat and will merge 😄

nsarrazin commented:
Works well, merging it! thanks again for an awesome contrib @mikelfried
